compiler/src/dotty/tools/dotc/semanticdb/TypeOps.scala

```diff
@@ -0,0 +1,452 @@
+            // instead of `("y", outer) -> <y>`
+            if lam.paramNames.contains(sym.name) then
+              paramRefSymtab((lam, sym.name)) = sym
+            else
+              enterParamRef(lam.resType)
+
+          // for class constructor
+          // class C[T] { ... }
+          case cls: ClassInfo if sym.info.isInstanceOf[LambdaType] =>
+            val lam = sym.info.asInstanceOf[LambdaType]
+            cls.cls.typeParams.foreach { param =>
+              paramRefSymtab((lam, param.name)) = param
+            }
+
+          // type X[T] = ...
+          case tb: TypeBounds =>
+            enterParamRef(tb.lo)
+            enterParamRef(tb.hi)
+
+          case _ => ()
+        }
+
+      def enterRefined(tpe: Type): Unit =
+        tpe match {
+          case refined: RefinedType =>
+            val key = (refined, sym.name)
+            refinementSymtab(key) = sym
+
+          case rec: RecType =>
+            enterRefined(rec.parent)
+
+          // Register symbol for opaque type,
+          // opaque type alias will be stored into the refinement of
+          // the self type of the enclosing class.
+          // Key: the tuple of
+          //   - self-type of enclosing class
+          //   - name of the opaque type
+          // Value: the symbol of the opaque type
+          // See: SymDenotation.opaqueToBounds
+          case cls: ClassInfo if sym.is(Flags.Opaque) =>
+            cls.classSymbol.asClass.givenSelfType match
+              case rt: RefinedType =>
+                refinementSymtab((rt, sym.name)) = sym
+              case _ => ()
+
+          case cls: ClassInfo if (cls.cls.name == tpnme.REFINE_CLASS) =>
+            enterRefined(sym.owner.owner.info)
+
+          // type x = Person { refinement }
+          case tb: TypeBounds =>
+            // tb = TypeBounds(
+            //   lo = RefinedType(...)
+            //   hi = RefinedType(...)
+            // )
+            enterRefined(tb.lo)
+            enterRefined(tb.hi)
+
+          // def s(x: Int): { refinement } = ...
+          case expr: ExprType =>
+            enterRefined(expr.resType)
+          case m: LambdaType =>
+            enterRefined(m.resType)
+          case _ => ()
+        }
+      if sym.exists && sym.owner.exists then
+        enterParamRef(sym.owner.info)
+        enterRefined(sym.owner.info)
+
+      def loop(tpe: Type): s.Signature = tpe match {
+        case mp: MethodOrPoly =>
+          def flatten(
+            t: Type,
+            paramss: List[List[SemanticSymbol]],
+            tparams: List[SemanticSymbol]
+          ): (Type, List[List[SemanticSymbol]], List[SemanticSymbol]) = t match {
+            case mt: MethodType =>
+              val syms: List[SemanticSymbol] = mt.paramNames.zip(mt.paramInfos).map { (name, info) =>
+                paramRefSymtab.lookup(mt, name).getOrElse {
+                  val fakeSym = TermParamRefSymbol(sym, name, info)
```

Comment:
could make use of `import scala.util.chaining.*`
```suggestion
                  TermParamRefSymbol(sym, name, info).tap(registerFakeSymbol)
```